### DISCUSSION TITLE
Support to create bucket with special 'num_shards'?

### DISCUSSION THREAD

**Discussion by zhucan** [rook/rook#12565] [Ideas]
![image](https://github.com/rook/rook/assets/13997135/3388a54c-05b8-42b1-b0b2-e864f47f93e5)

**Comment by travisn** [comment]
Have you tried the ceph.conf override for `rgw_override_bucket_index_max_shards`? From that code snippet it looks like that would work.

**Reply by zhucan** [reply]
we want diffirent buckets with different num_shards.

**Reply by travisn** [reply]
Then there are two questions to investigate:
1. How is num_shards applied to an rgw bucket? 
2. How are you creating and managing the bucket? OBCs or COSI?

**Reply by zhucan** [reply]
Because of the dynamic of shards  will effect the w IO, so we need to assign a special num shards. We create the bucket with OBCs.

**Comment by BlaineEXE** [comment]
Given that OBCs are (from the library Rook uses to provide the support) deprecated and since COSI is the replacement. I think it is important to consider this request in light of the COSI design. In the COSI design, this information is not part of the broader object/bucket design and is driver specific (i.e., Ceph-specific). Any driver specific config can only be on the BucketClass. If this feature is desired, the only implementation we can accomplish in Rook is to support the parameter there. When end users need a special number of shards, they can use the BucketClass (that they have RBAC permissions to use) that most closely meets their performance needs. That is the only possible implementation for COSI.

With that in mind, I think that is the also the implementation we should follow for OBCs as well. At this point, we would be doing ourselves and our users a disservice if we made the OBC implementation divergent from COSI. OBCs **will** be removed from Rook support eventually.

**Reply by BlaineEXE** [reply]
As a follow up, is it possible to set `num_shards` on an OBC StorageClass today to accomplish this?

**Reply by zhucan** [reply]
@BlaineEXE sorry for delayed reponses, I think the **num_shards** on OBC is suitable more than on sc.

**Reply by BlaineEXE** [reply]
num_shards is far too technical of a field to expose to each user. Exposing it to self-service users who are not well informed to cause Ceph cluster health issues. 

Further, it will be impossible to allow this for COSI as explained above. To ensure that OBC users will have the most direct migration path to COSI, Rook should not implement this on a per-OBC basis.

Because of these two very clear reasons, Rook will not implement this on the OBC itself.

Storage administrators who want their self-service users to be able to control num_shards could reasonably be allowed to set this in the StorageClass for OBCs. However, I don't know offhand whether it is currently possible for `num_shards` to be configured via OBC StorageClasses today. If not, that functionality can be added via the opaque parameters.
